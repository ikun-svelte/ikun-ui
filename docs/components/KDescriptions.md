---
title: KDescriptions
lang: en-US
---

# KDescriptions

Display multiple fields in list form.

## Install

::: code-group

```bash [pnpm]
pnpm add @ikun-ui/descriptions @ikun-ui/descriptions-item
```

```bash [yarn]
yarn add @ikun-ui/descriptions @ikun-ui/descriptions-item
```

```bash [npm]
npm install @ikun-ui/descriptions @ikun-ui/descriptions-item
```

:::

## Basic usage

<demo src="descriptions/basic.svelte"  github='Descriptions'></demo>

## Sizes

<demo src="descriptions/sizes.svelte"  github='Descriptions'></demo>

## Vertical

<demo src="descriptions/vertical.svelte"  github='Descriptions'></demo>

## Descriptions Props

| Name      | Type                      | Default      | Description                                |
| --------- | ------------------------- | ------------ | ------------------------------------------ |
| title     | `string`                  | `''`         | title text, display on the top left        |
| extra     | `string`                  | `''`         | extra text, display on the top right       |
| border    | `boolean`                 | `false`      | with or without border                     |
| column    | `number`                  | `3`          | numbers of `Descriptions Item` in one line |
| direction | `horizontal` / `vertical` | `horizontal` | direction of descriptions                  |
| size      | `sm` / `md` / `lg`        | `md`         | size of descriptions                       |
| cls       | `string`                  | `''`         | Additional class for component             |
| attrs     | `Record<string, string>`  | `{}`         | Additional attributes                      |

## Descriptions Slots

| Name    | Description               |
| ------- | ------------------------- |
| title   | Customize title content   |
| extra   | Customize extra content   |
| default | Customize default content |

## DescriptionsItem Props

| Name  | Type                     | Default | Description                    |
| ----- | ------------------------ | ------- | ------------------------------ |
| label | `string`                 | `''`    | label text                     |
| cls   | `string`                 | `''`    | Additional class for component |
| attrs | `Record<string, string>` | `{}`    | Additional attributes          |

## DescriptionsItem Slots

| Name    | Description               |
| ------- | ------------------------- |
| label   | Customize label content   |
| default | Customize default content |
