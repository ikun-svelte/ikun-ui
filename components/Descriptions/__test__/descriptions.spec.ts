import { tick } from 'svelte';
import { afterEach, expect, test, describe, beforeEach } from 'vitest';
import KDescriptions from '../src';
import KDescriptionsDirection from './fixture/descriptions.direction.svelte';
import KDescriptionsTitle from './fixture/descriptions.title.svelte';
import KDescriptionsExtra from './fixture/descriptions.extra.svelte';

let host: HTMLElement;

const initHost = () => {
	host = document.createElement('div');
	host.setAttribute('id', 'host');
	document.body.appendChild(host);
};
beforeEach(() => {
	initHost();
});
afterEach(() => {
	host.remove();
});

describe('Test: KDescriptions', () => {
	test('props: cls', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				cls: 'k-descriptions--test'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--test')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: title', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				title: 'KDescriptions Title'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--title')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('KDescriptions Title')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: extra', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				extra: 'KDescriptions Extra'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--extra')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('KDescriptions Extra')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: column', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				column: 5
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--cols-5')).toBeTruthy();
		instance.$set({ column: 3 });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--cols-3')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: border', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				border: false
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--body--border')).toBeFalsy();
		instance.$set({ border: true });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--body--border')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: sizes', async () => {
		const instance = new KDescriptions({
			target: host,
			props: {
				title: 'KDescriptions',
				size: 'sm'
			}
		});
		expect(instance).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--sm')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--header--sm')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--body--sm')).toBeTruthy();
		instance.$set({ size: 'md' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--md')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--header--md')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--body--md')).toBeTruthy();
		instance.$set({ size: 'lg' });
		await tick();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--lg')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--header--lg')).toBeTruthy();
		expect((host as HTMLElement)!.innerHTML.includes('k-descriptions--body--lg')).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('props: direction', async () => {
		const instance = new KDescriptionsDirection({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		expect(host.querySelectorAll('.k-descriptions-item__horizontal').length === 3).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
		const btnEl = host.querySelector('button');
		btnEl?.click();
		await tick();
		expect(host.querySelectorAll('.k-descriptions-item__vertical').length === 3).toBeTruthy();
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: title', async () => {
		const instance = new KDescriptionsTitle({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const titleEl = host.querySelector('#slot-title');
		expect(titleEl!.textContent).toBe('Ikun Descriptions Title');
		expect(host.innerHTML).matchSnapshot();
	});

	test('slots: extra', async () => {
		const instance = new KDescriptionsExtra({
			target: host
		});
		expect(instance).toBeTruthy();
		await tick();
		const extraEl = host.querySelector('#slot-extra');
		expect(extraEl!.textContent).toBe('Ikun Descriptions Extra');
		expect(host.innerHTML).matchSnapshot();
	});
});
