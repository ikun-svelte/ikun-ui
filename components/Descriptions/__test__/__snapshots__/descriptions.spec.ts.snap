// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Test: KDescriptions > props: border 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"> <div class=\\"k-descriptions--body k-descriptions--cols-3 k-descriptions--body--border\\"></div></div>"`;

exports[`Test: KDescriptions > props: cls 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark k-descriptions--test\\"> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > props: column 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > props: direction 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"><div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__horizontal ikun-last:k-descriptions-item--col-1\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__horizontal k-descriptions-item--label__horizontal--md\\">name</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__horizontal\\">IKun</div></div> <div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__horizontal ikun-last:k-descriptions-item--col-2\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__horizontal k-descriptions-item--label__horizontal--md\\">phone</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__horizontal\\">18868686868</div></div> <div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__horizontal ikun-last:k-descriptions-item--col-0\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__horizontal k-descriptions-item--label__horizontal--md\\">age</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__horizontal\\">1</div></div></div></div> <div id=\\"show-direction\\">horizontal</div> <button>change direction</button>"`;

exports[`Test: KDescriptions > props: direction 2`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"><div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__vertical ikun-last:k-descriptions-item--col-1\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__vertical k-descriptions-item--label__vertical--md\\">name</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__vertical\\">IKun</div></div> <div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__vertical ikun-last:k-descriptions-item--col-2\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__vertical k-descriptions-item--label__vertical--md\\">phone</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__vertical\\">18868686868</div></div> <div class=\\"k-descriptions-item k-descriptions-item__dark k-descriptions-item--md k-descriptions-item__vertical ikun-last:k-descriptions-item--col-0\\"><label class=\\"k-descriptions-item--label k-descriptions-item--label__vertical k-descriptions-item--label__vertical--md\\">age</label> <div class=\\"k-descriptions-item--content k-descriptions-item--content__vertical\\">1</div></div></div></div> <div id=\\"show-direction\\">vertical</div> <button>change direction</button>"`;

exports[`Test: KDescriptions > props: extra 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"><div class=\\"k-descriptions--header k-descriptions--header--md\\"><div class=\\"k-descriptions--title\\"></div> <div class=\\"k-descriptions--extra\\">KDescriptions Extra</div></div> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > props: sizes 1`] = `"<div class=\\"k-descriptions k-descriptions--lg k-descriptions__dark\\"><div class=\\"k-descriptions--header k-descriptions--header--lg\\"><div class=\\"k-descriptions--title\\">KDescriptions</div> <div class=\\"k-descriptions--extra\\"></div></div> <div class=\\"k-descriptions--body k-descriptions--body--lg k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > props: title 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"><div class=\\"k-descriptions--header k-descriptions--header--md\\"><div class=\\"k-descriptions--title\\">KDescriptions Title</div> <div class=\\"k-descriptions--extra\\"></div></div> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > slots: extra 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"><div class=\\"k-descriptions--header k-descriptions--header--md\\"><div class=\\"k-descriptions--title\\"></div> <div class=\\"k-descriptions--extra\\"><div id=\\"slot-extra\\" slot=\\"extra\\">Ikun Descriptions Extra</div></div></div> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;

exports[`Test: KDescriptions > slots: title 1`] = `"<div class=\\"k-descriptions k-descriptions--md k-descriptions__dark\\"><div class=\\"k-descriptions--header k-descriptions--header--md\\"><div class=\\"k-descriptions--title\\"><div id=\\"slot-title\\" slot=\\"title\\">Ikun Descriptions Title</div></div> <div class=\\"k-descriptions--extra\\"></div></div> <div class=\\"k-descriptions--body k-descriptions--body--md k-descriptions--cols-3\\"></div></div>"`;
